src/ChebyshevApproximationInterface.cc

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC and
+// UT-Battelle, LLC.
+// Produced at the Lawrence Livermore National Laboratory and the Oak Ridge
+// National Laboratory.
+// LLNL-CODE-743438
+// All rights reserved.
+// This file is part of MGmol. For details, see https://github.com/llnl/mgmol.
+// Please also read this link https://github.com/llnl/mgmol/LICENSE
+
+#include "ChebyshevApproximationInterface.h"
+
+Timer ChebyshevApproximationInterface::compute_coeffs_tm_(
+    "ChebyshevApproximationInterface::compChebCoeffs");
+Timer ChebyshevApproximationInterface::double_loop_tm_(
+    "ChebyshevApproximationInterface::double_loop");
+
+void ChebyshevApproximationInterface::evaluateFunction(
+    const std::vector<double>& fnodes, std::vector<double>& fvals)
+{
+    fvals.clear();
+    fvals = chebfunc_->eval(fnodes);
+}
+
+void ChebyshevApproximationInterface::computeChebyshevCoeffs()
+{
+    compute_coeffs_tm_.start();
+    order_ = max_order_;
+    int n  = order_;
+    // define scaling variable here
+    const double fac = 2.0 / static_cast<double>(n);
+    // evaluate function at interpolation points
+    std::vector<double> fvals;
+    evaluateFunction(interp_points_, fvals);
+
+    // build coefficients
+    coeffs_.clear();
+    coeffs_.reserve(order_);
+
+    // initialize first entry of coeffs_
+    double fsum = 0.;
+    for (const double& value : fvals)
+        fsum += value;
+    // scale and store first coeff
+    fsum /= static_cast<double>(n);
+    coeffs_.push_back(fsum);
+
+    // compute remaining coeffs_ values
+    double_loop_tm_.start();
+
+    LocalVector<double> fvec(fvals);
+    coeffs_.resize(n, 0.);
+    LocalVector<double> cvec(coeffs_);
+
+    cmat_->matvec(fvec, cvec);
+    (cvec.data())[0] = fsum;
+    cvec.swap(coeffs_);
+
+    double_loop_tm_.stop();
+
+    // scale only last n-1 coefficients
+    const int n1 = n - 1;
+    MPscal(n1, fac, &coeffs_[1]);
+
+    //    for (int i = 1; i < n; i++)assert(coeffs_[i] == coeffs_[i]);
+
+    compute_coeffs_tm_.stop();
+}
+
+// Compute Chebyshev approximation given function points.
+// Assume that points are sorted from left to right on the real axis.
+// Example: (sorted) eigenvalues of some matrix.
+void ChebyshevApproximationInterface::computeChebyshevApproximation(
+    std::vector<double>& points, std::vector<double>& vals)
+{
+    assert(static_cast<int>(points.size()) == order_);
+    vals.clear();
+    // get extents of points
+    const double a = points[0];
+    const double b = points[order_ - 1];
```

Comment:
There is a warning that `a` and `b` are not used.

Reply:
Fixed thanks. Question, do we not use -Wall flags by default in our build? I did not see any such warning when I built the code before pushing the last time.

Reply:
No, not by default. Also the CI is using -Wall -pedantic -Wextra

Reply:
Ok, good to know. Thanks